stablehlo/tests/infer_stablehlo.mlir

```diff
@@ -992,6 +992,36 @@ func.func @concat_bounds_unranked_c1(
 
 // -----
 
+// CHECK-LABEL: while_bounds
+func.func @while_bounds(
+  %while_arg_1: tensor<1xi32>,
+  %while_arg_2: tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>,
+  %while_arg_3: tensor<1xf32>,
+  %while_arg_4: tensor<3xf32>) -> tensor<*xindex> {
+  %1:4 = "stablehlo.while"(%while_arg_1, %while_arg_2, %while_arg_3, %while_arg_4) ({
+  ^bb0(%arg1: tensor<1xi32>, %arg2: tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>, %arg3: tensor<1xf32>, %arg4: tensor<3xf32>):
+    %2 = arith.constant dense<0> : tensor<i32>
+    %3 = "stablehlo.slice"(%arg2) {limit_indices = dense<[1, 1]> : tensor<2xi64>, start_indices = dense<[0, 0]> : tensor<2xi64>, strides = dense<[1, 1]> : tensor<2xi64>} :
+      (tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>) -> tensor<1x1xi32>
+    %4 = "stablehlo.compare"(%3, %3) {comparison_direction = #stablehlo<comparison_direction LT>} : (tensor<1x1xi32>, tensor<1x1xi32>) -> tensor<1x1xi1>
+    %5 = "stablehlo.reshape"(%4) : (tensor<1x1xi1>) -> tensor<i1>
+    "stablehlo.return"(%5) : (tensor<i1>) -> ()
+  },  {
+  ^bb0(%arg1: tensor<1xi32>, %arg2: tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>, %arg3: tensor<1xf32>, %arg4: tensor<3xf32>):
+    %3 = "stablehlo.broadcast_in_dim"(%arg3) {broadcast_dimensions = dense<0> : tensor<1xi64>} : (tensor<1xf32>) -> tensor<3xf32>
+    %4 = stablehlo.add %3, %arg4 : tensor<3xf32>
+    "stablehlo.return"(%arg1, %arg2, %arg3, %4) : (tensor<1xi32>, tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>, tensor<1xf32>, tensor<3xf32>) -> ()
+  }) : (tensor<1xi32>, tensor<2x?xi32, #stablehlo.type_extensions<bounds = [?, 4]>>, tensor<1xf32>, tensor<3xf32>) -> (tensor<*xi32>, tensor<*xi32>, tensor<*xf32>, tensor<*xf32>)
```

Comment:
How about simplifying the while loop as much as possible? E.g. the cond could be replaced with `return true`, and the body could be replaced with `return arguments`. Similarly, do we need all the 4 operands for this while op? types0, types2, types3 test exactly the same thing as far as this test is concerned.

Reply:
Good suggestion! Fixed.
I keep 2 args: 1 with bounds and 1 not, for testing that bounds and non-bounds can exist together in this multiple case.